query-algebrizer/src/clauses/not.rs

```diff
@@ -0,0 +1,534 @@
+        let d1v = QualifiedAlias::new(d1.clone(), DatomsColumn::Value);
+
+        let mut subquery = ConjoiningClauses::default();
+        subquery.from = vec![SourceAlias(DatomsTable::Datoms, d1)];
+        subquery.column_bindings.insert(vx.clone(), vec![d0e.clone(), d1e.clone()]);
+        subquery.wheres = ColumnIntersection(vec![ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d1a.clone(), knows.clone())),
+                                                  ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d1v.clone(), john)),
+                                                  ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d0e.clone(), QueryValue::Column(d1e.clone())))]);
+
+        subquery.known_types.insert(vx.clone(), ValueTypeSet::of_one(ValueType::Ref));
+        subquery.known_types.insert(vy.clone(), ValueTypeSet::of_one(ValueType::String));
+
+        let mut input_vars: BTreeSet<Variable> = BTreeSet::default();
+        input_vars.insert(vy.clone());
+        subquery.input_variables = input_vars;
+        subquery.value_bindings.insert(vy.clone(), TypedValue::typed_string("John"));
+
+        assert!(!cc.is_known_empty());
+        assert_eq!(cc.wheres, ColumnIntersection(vec![
+                ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d0a.clone(), knows)),
+                ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d0v.clone(), bill)),
+                ColumnConstraintOrAlternation::Constraint(ColumnConstraint::NotExists(ComputedTable::Subquery(subquery))),
+            ]));
+    }
+
+    // Test that if any single clause in the `not` fails to resolve the whole clause is considered empty
+    #[test]
+    fn test_fails_if_any_clause_invalid() {
+        let schema = prepopulated_schema();
+        let query = r#"
+            [:find ?x
+             :where [?x :foo/knows "Bill"]
+                    (not [?x :foo/nope "John"]
+                         [?x :foo/parent "Ámbar"]
+                         [?x :foo/nope "Daphne"])]"#;
+        let cc = alg(&schema, query);
+        assert!(!cc.is_known_empty());
+        compare_ccs(cc,
+                    alg(&schema,
+                        r#"[:find ?x :where [?x :foo/knows "Bill"]]"#));
+    }
+
+    /// Test that if all the attributes in an `not` fail to resolve, the `cc` isn't considered empty.
+    #[test]
+    fn test_no_clauses_succeed() {
+        let schema = prepopulated_schema();
+        let query = r#"
+            [:find ?x
+             :where [?x :foo/knows "John"]
+                    (not [?x :foo/nope "Ámbar"]
+                         [?x :foo/nope "Daphne"])]"#;
+        let cc = alg(&schema, query);
+        assert!(!cc.is_known_empty());
+        compare_ccs(cc,
+                    alg(&schema, r#"[:find ?x :where [?x :foo/knows "John"]]"#));
+                    
+    }
+
+    #[test]
+    #[should_panic]
```

Comment:
Let's not use `alg`; algebrize it directly so you can use `assert!(….is_err())`. We want to make sure that a bad query like this doesn't panic, but fails, and this test hides that distinction.